src/Cardano/Address/Style/Shelley.hs

```diff
@@ -0,0 +1,217 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE InstanceSigs #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+{-# OPTIONS_HADDOCK prune #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+
+module Cardano.Address.Style.Shelley
+    ( -- $overview
+
+      -- * Shelley
+      Shelley
+
+      -- * Accessors
+    , getKey
+
+      -- * Unsafe
+    , liftXPrv
+
+      -- Internals
+    , minSeedLengthBytes
+    , publicKeySize
+    , addrSingleSize
+    , addrGroupedSize
+    ) where
+
+import Prelude
+
+import Cardano.Address.Derivation
+    ( AccountingStyle
+    , Depth (..)
+    , DerivationScheme (..)
+    , DerivationType (..)
+    , GenMasterKey (..)
+    , HardDerivation (..)
+    , Index
+    , SoftDerivation (..)
+    , XPrv
+    , deriveXPrv
+    , deriveXPub
+    , generateNew
+    )
+import Cardano.Mnemonic
+    ( SomeMnemonic (..), entropyToBytes, mnemonicToEntropy )
+import Control.DeepSeq
+    ( NFData )
+import Control.Exception.Base
+    ( assert )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Word
+    ( Word32 )
+import GHC.Generics
+    ( Generic )
+
+import qualified Data.ByteArray as BA
+
+-- $overview
+--
+-- This module provides an implementation of:
+--
+-- - 'GenMasterKey': for generating Shelley master keys from mnemonic sentences
+-- - 'HardDerivation': for hierarchical hard derivation of parent to child keys
+-- - 'SoftDerivation': for hierarchical soft derivation of parent to child keys
+--
+
+{-------------------------------------------------------------------------------
+                                   Key Types
+-------------------------------------------------------------------------------}
+-- | A cryptographic key for sequential-scheme address derivation, with
+-- phantom-types to disambiguate key types.
+--
+-- @
+-- let rootPrivateKey = Shelley 'RootK XPrv
+-- let accountPubKey  = Shelley 'AccountK XPub
+-- let addressPubKey  = Shelley 'AddressK XPub
+-- @
+--
+-- @since 1.0.0
+newtype Shelley (depth :: Depth) key = Shelley
+    { getKey :: key
+        -- ^ Extract the raw 'XPrv' or 'XPub' wrapped by this type.
+        --
+        -- @since 1.0.0
+    }
+    deriving stock (Generic, Show, Eq)
+
+deriving instance (Functor (Shelley depth))
+instance (NFData key) => NFData (Shelley depth key)
+
+-- | Unsafe backdoor for constructing an 'Shelley' key from a raw 'XPrv'. this is
+-- unsafe because it lets the caller choose the actually derivation 'depth'.
+--
+-- This can be useful however when serializing / deserializing such a type, or to
+-- speed up test code (and avoid having to do needless derivations from a master
+-- key down to an address key for instance).
+--
+-- @since 1.0.0
+liftXPrv :: XPrv -> Shelley depth XPrv
+liftXPrv = Shelley
+
+instance GenMasterKey Shelley where
+    type SecondFactor Shelley = Maybe SomeMnemonic
```

Comment:
SecondFactor should just be a plain `ScrubbedBytes` here. cardano-wallet's API may chose to encode this as a mnemonic, but that's only from an API perspective and shouldn't leak inside this module.